native/src/main/scala/sqs4s/auth/CredProvider.scala

```diff
@@ -0,0 +1,159 @@
+package sqs4s.auth
+
+import java.util.concurrent.TimeUnit
+
+import cats.Applicative
+import cats.implicits._
+import cats.effect.implicits._
+import cats.effect.concurrent.Ref
+import cats.effect.{Clock, Concurrent, Sync, Timer}
+import fs2._
+import org.http4s.Method.{GET, PUT}
+import org.http4s.circe.CirceEntityDecoder._
+import org.http4s.client.Client
+import org.http4s.client.dsl.Http4sClientDsl
+import org.http4s.implicits._
+import org.http4s.{Header, Response, Status}
+import sqs4s.auth.errors._
+
+import scala.concurrent.duration._
+
+trait CredProvider[F[_]] {
+  def accessKey: F[String]
+  def secretKey: F[String]
+}
+
+trait SessionCredProfiler[F[_]] {
+  def sessionToken: F[String]
+}
+
+trait TemporaryCredProvider[F[_]]
+    extends CredProvider[F]
+    with SessionCredProfiler[F] {
+  def refresh: F[Credential]
+}
+
+case class BasicCredProvider[F[_]: Applicative](
+  accessK: String,
+  secretK: String
+) extends CredProvider[F] {
+  def accessKey: F[String] = accessK.pure[F]
+  def secretKey: F[String] = secretK.pure[F]
+}
+
+class InstanceProfileCredProvider[F[_]: Sync: Clock: Timer](
+  client: Client[F],
+  ref: Ref[F, Option[Credential]],
+  ttl: FiniteDuration,
+  refreshBefore: FiniteDuration
+) extends TemporaryCredProvider[F]
+    with Http4sClientDsl[F] {
+
+  private val linkLocal = uri"http://169.254.169.254"
```

Comment:
What is this address for?

Reply:
it is the address to retrieve cred from per doc [here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials) and [AWS SDK](https://github.com/aws/aws-sdk-java/blob/542606ddd5ce6f0c3570ef7488c4f5479c6425bb/aws-java-sdk-core/src/main/java/com/amazonaws/util/EC2MetadataUtils.java#L73)

Reply:
Oh, I didn't noticed before they were using a static IP instead of some DNS, that suprises me.

I would:

1. Rename it as `awsLocalLink`, mostly because the name _local_ made me think it was for testing and that you forgot to remove it.
2. Add your response as a comment in the code, since I am sure any other developer reading the code would want an explanation of why having a hardcoded IP address.